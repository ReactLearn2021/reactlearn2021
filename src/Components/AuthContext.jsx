import React, { useState } from "react";

export const AuthContext = React.createContext();
export const AuthProvider = ({ children }) => {
    const [loggedIn, setLoggedIn] = useState(false),
          logIn = (email, password) => {
                if (email !== process.env.REACT_APP_ACCESS_EMAIL || password !== process.env.REACT_APP_ACCESS_PASSWORD) {
                    return;
                }

                setLoggedIn(true);
          },
          logOut = () => {
                setLoggedIn(false);
          };

    return(
        <AuthContext.Provider value = {{ logIn, logOut, loggedIn }}>
            { children }
        </AuthContext.Provider>
    )
};

export const withAuth = (Wrapped) => {
    return class extends React.Component {
        render() {
            return(
                <AuthContext.Consumer>
                    {
                        (value) => <Wrapped {...value} {...this.props} />
                    }
                </AuthContext.Consumer>
            )
        }
    }
}