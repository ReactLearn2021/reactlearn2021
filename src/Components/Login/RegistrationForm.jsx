import React from "react";
import classNames from "classnames";
import propTypes from "prop-types";
import { connect } from "react-redux";
import { REGISTER } from "../../store/actions";
import { Formik, ErrorMessage, Form, Field } from "formik";

export default class RegistrationForm extends React.Component {
    constructor(props) {
        super(props);
        this.handleChangeField = this.handleChangeField.bind(this);
        this.regHandler = this.regHandler.bind(this);
    }

    static propTypes = {
        children : propTypes.node,
        register : propTypes.func
    }

    state = {
        regFields : {
            email : "",
            initials : "",
            setPassword : ""
        },
        emailCheck : /^([\w\d-\.\?]+)@[A-Z]{4,6}\.[A-Z]{2,4}$/i
    };

    handleChangeField(event) {
        this.setState({ [event.target.name]: event.target.value });
    }

    regHandler() {
        this.props.register(this.state.regFields.email, this.state.regFields.initials, this.state.regFields.setPassword);
    }

    render() {
        return(
                <Formik onSubmit = { (values) => { return this.regHandler(values) } } 
                initialValues = { { email : "", initials : "", setPassword : "" } }
                validate = { (values) => {
                    const errors = {};
                    if (!this.state.emailCheck.test(values.email)) {
                        errors.login = "Введен некоректный email";
                    }

                    if (values.setPassword.length < 8) {
                        errors.password = "Введен некоректный пароль";
                    }

                    if (values.initials.length < 1) {
                        errors.initials = "Введено некорретное имя пользователя";
                    }

                    return errors;
                } }
                >{ ({ handleChange, handleBlur, values, errors }) => {
                    const btnClass = classNames({
                        "loft__form-button" : Object.keys(errors).length > 0 || Object.values(values).some( (item) => item === ""),
                        "loft__form-button-filled" : !errors || Object.keys(errors).length === 0
                    });
                    return(
                    <Form className = "authorize__block-form">
                        <h2>Регистрация</h2>
                        <Field type = "text" 
                        name = "email" 
                        id = "email" 
                        data-testid = "email" 
                        className = "loft__form-input"
                        autoComplete = "off"
                        onBlur = { handleBlur }
                        onChange = { handleChange } 
                        value = { values.email } />
                        <label htmlFor = "email">Адрес электронной почты <sup>&#10057;</sup></label>
                        <ErrorMessage name = "email" component = "p" className = "error-message" />
                        <Field type = "text" 
                        name = "initials" 
                        id = "initials" 
                        data-testid = "initials" 
                        className = "loft__form-input" 
                        autoComplete = "off"
                        onBlur = { handleBlur }
                        onChange = { handleChange }
                        value = { values.initials } />
                        <label htmlFor = "initials">Как вас зовут? <sup>&#10057;</sup></label>
                        <ErrorMessage name = "initials" component = "p" className = "error-message" />
                        <Field type = "text" 
                        name = "setPassword" 
                        id = "setPassword" 
                        data-testid = "setPassword" 
                        className = "loft__form-input" 
                        autoComplete = "off"
                        onBlur = { handleBlur }
                        onChange = { handleChange }
                        value = { values.setPassword } />
                        <label htmlFor = "setPassword">Придумайте пароль <sup>&#10057;</sup></label>
                        <ErrorMessage name = "setPassword" component = "p" className = "error-message" />
                        <button className = { btnClass } type = "submit" data-testid = "reg-button" disabled = { (btnClass === "loft__form-button") ? true : false }>Зарегистрироваться</button>
                        <div className = "loft__form-anchor">
                            { this.props.children }
                        </div>
                    </Form>
                );
            }} 
            </Formik>               
        )
    }
}

export const RegisterFormWithAuth = connect(
    null,
    { register : REGISTER }
)(RegistrationForm);