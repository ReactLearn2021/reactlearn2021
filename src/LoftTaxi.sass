// body
//     overflow: hidden

@font-face
    font-family: Lora
    src: url(/assets/fonts/Lora-VariableFont_wght.ttf)

body
    margin: 0
    font-family: Lora
    -webkit-font-smoothing: antialiased
    -moz-osx-font-smoothing: grayscale

*
    outline: none

code 
    font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New', monospace

html, body, #root, section, main
    height: 100%

.left__logo-image
    height: 100%
    width: 34%
    background: #1C1A19
    display: flex
    flex-flow: row nowrap
    align-items: center
    justify-content: center

    img
        margin-right: 6%
        -webkit-user-select: none
        user-select: none

.auth__section
    background-image: url("./assets/auth-image.jpg")
    background-size: cover
    height: 100%
    width: 66%
    display: flex
    align-items: center
    justify-content: center

.authorize__block
    background: white
    padding: 46px 0
    box-shadow: 0px 0px 40px 6px rgba(0, 0, 0, 0.1)
    border-radius: 22px

.authorize__block-form
    padding: 0 102px 0 98px
    display: flex
    flex-flow: column nowrap
    justify-content: center

    h2
        text-align: center
        margin-bottom: 50px

sup
    color: firebrick

label
    position: relative
    bottom: 31px
    left: 20px
    transition: transform 200ms ease-out
    cursor: text

.loft__form-input
    width: 300px
    margin-left: 20px
    height: 35px
    border: none
    border-bottom: 1px solid black !important
    font-size: 16px
    margin-top: 20px

    &:focus + label
        transform: translateY(-31px) translateX(-34px) scale(0.8)
        transition: transform 200ms ease-out

    &:not([value=""]) + label
         transform: translateY(-31px) translateX(-34px) scale(0.8) !important

.loft__form-button
    font-size: 20px
    background-color: #D8D7D5
    color: #737373
    margin-top: 30px
    padding: 12px 15px
    border-radius: 40px
    border: none
    cursor: pointer

    &+div
        margin-top: 30px
        font-size: 17px
        
        a
            color: rgb(253, 191, 90)
            text-decoration: none

    &-filled
        background: #FDBF5A
        color: black

        &:focus
            background: #FFA842

.loft__form-anchor
    text-align: center
    width: 350px