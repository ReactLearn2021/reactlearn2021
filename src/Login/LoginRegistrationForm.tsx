import React, { ReactElement, ReactNode } from "react";

const enum Views {
    LOGIN = "LOGIN",
    REG = "REG"
}

export default class LoginRegistrationForm extends React.Component {

    constructor(public props: Readonly<{}> & Readonly<{ children?: ReactNode; }>) {
        super(props);
        this.handleChangeField = this.handleChangeField.bind(this);
        this.loginHandler = this.loginHandler.bind(this);
        this.changeCurrentView = this.changeCurrentView.bind(this);
        this.regHandler = this.regHandler.bind(this);
    }

    public state = {
        login : "",
        password : "",
        email : "",
        initials : "",
        setPassword : "",
        fullLoginForm : false,
        fullRegForm : false,
        loginButton : "loft__form-button",
        regButton : "loft__form-button",
        currentView : Views.LOGIN,
        emailCheck : /^([\w\d-\.\?]+)@[A-Z]{4,6}\.[A-Z]{2,4}$/i
    };

    public changeCurrentView(param: string): void {
        if (this.state.currentView == Views.LOGIN) {
            this.setState({ currentView : Views.REG });
        } else {
            this.setState({ currentView : Views.LOGIN });
        }
    }

    public handleChangeField(event: { target: { value: string; name: string }}): void {
        console.info(event.target.name)
        this.setState({ [event.target.name]: event.target.value });
    }

    public loginHandler(): boolean {
        if (!this.state.emailCheck.test(this.state.login)) {
            return false;
        }
        console.info(true);
        return true;
    }

    public regHandler(): boolean {
        if (this.state.setPassword.length <= 8) {
            return false;
        }
        return true;
    }

    public componentDidUpdate(): void {
        if (this.state.emailCheck.test(this.state.login) && this.state.password.length >= 8 && this.state.loginButton != "loft__form-button-filled") {
            this.setState({ loginButton : "loft__form-button-filled" });
        } else if (this.state.loginButton == "loft__form-button") {
            return;
        } else {
            this.setState({ loginButton : "loft__form-button" }); // спросить про это
        }
    }

    public render(): ReactElement {
        // const login = this.state.login;
        return(
            <div className = "authorize__block">
                {
                    {
                        LOGIN : <div className = "authorize__block-form">
                            <h2>Войти</h2>
                            <input type = "text" name = "login" id = "login" className = "loft__form-input" value = { this.state.login } onChange = { () => this.handleChangeField("Param") }/> {/* this нужен потому что классовый компонент */}
                            <label htmlFor = "login">Имя пользователя <sup>&#10057;</sup></label>
                            <input type = "password" name = "password" id = "password" className = "loft__form-input" value = { this.state.password } onChange = { this.handleChangeField }/> 
                            <label htmlFor = "password">Пароль <sup>&#10057;</sup></label>
                            <button className = { this.state.loginButton } type = "button" onClick = { this.loginHandler }>Войти</button>
                            <div className = "loft__form-anchor">
                                <span>Новый пользователь? <a href = "javascript:void(0)" onClick = { () => this.changeCurrentView("Param") }>Зарегистрироваться</a></span>
                            </div>
                        </div>
                        ,
                        REG : <div className = "authorize__block-form">
                            <h2>Регистрация</h2>
                            <input type = "text" name = "email" id = "email" className = "loft__form-input" value = { this.state.email } onChange = { this.handleChangeField }/>
                            <label htmlFor = "email">Адрес электронной почты <sup>&#10057;</sup></label>
                            <input type = "text" name = "initials" id = "initials" className = "loft__form-input" value = { this.state.initials } onChange = { this.handleChangeField }/>
                            <label htmlFor = "initials">Как вас зовут? <sup>&#10057;</sup></label>
                            <input type = "password" name = "setPassword" id = "setPassword" className = "loft__form-input" value = { this.state.setPassword } onChange = { this.handleChangeField }/>
                            <label htmlFor = "setPassword">Придумайте пароль <sup>&#10057;</sup></label>
                            <button className = { this.state.regButton } type = "button" onClick = { this.regHandler }>Зарегистрироваться</button>
                            <div className = "loft__form-anchor">
                                <span>Уже зарегистрированы? <a href = "javascript:void(0)" onClick = { () => this.changeCurrentView("Param") }>Войти</a></span>
                            </div>
                        </div>
                    }[this.state.currentView]
                }
            </div>
        );
    }

}