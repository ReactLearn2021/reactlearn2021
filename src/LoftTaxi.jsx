import './LoftTaxi.sass';
import LoginPage from "./Components/Login/LoginPage";
import React from "react";
import { MapWithAuth } from "./Components/Map/Map";
import { withAuth } from "./Components/AuthContext";
import propTypes from "prop-types";

function App(props) {
  // if (window.localStorage.getItem("view") != "map") {
    if (!props.loggedIn) {
        return (
          <main data-testid = "main">
              <LoginPage/>
          </main>
        );
    } else {
      return <main data-testid = "map-wrapper">
                  <MapWithAuth {...props} />
             </main>
    }
}

App.propTypes = {
    loggedIn : propTypes.bool,
    logIn : propTypes.func,
    logOut : propTypes.func
};

export default withAuth(App);
