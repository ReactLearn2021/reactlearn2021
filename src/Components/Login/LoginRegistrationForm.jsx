import React from "react";
import RegistrationForm from "./RegistrationForm";
import { LoginFormWithAuth } from "./LoginForm";

const Views = {
    LOGIN : "LOGIN",
    REG : "REG"
}

export default class LoginRegistrationForm extends React.Component {

    constructor(props) {
        super(props);
        this.changeCurrentView = this.changeCurrentView.bind(this);
    }

    state = {
        currentView : Views.LOGIN,
    };

    changeCurrentView(param) {
        if (this.state.currentView === Views.LOGIN) {
            this.setState({ currentView : Views.REG });
        } else {
            this.setState({ currentView : Views.LOGIN });
        }
    }

    render() {
        // const login = this.state.login;
        return(
            <div className = "authorize__block">
                {
                    {
                        LOGIN : <LoginFormWithAuth {...this.props}><span>Новый пользователь? <a onClick = { () => this.changeCurrentView("Param") }>Зарегистрироваться</a></span></LoginFormWithAuth>,
                        REG : <RegistrationForm><span>Уже зарегистрированы? <a onClick = { () => this.changeCurrentView("Param") }>Войти</a></span></RegistrationForm>
                    }[this.state.currentView]
                }
            </div>
        );
    }

}