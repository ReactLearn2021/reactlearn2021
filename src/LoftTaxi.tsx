import './LoftTaxi.sass';
import { ReactElement } from "react";
import LoginPage from "./Login/LoginPage";

function App(): ReactElement {
  return (
    <main>
        <LoginPage/>
    </main>
  );
}

export default App;
