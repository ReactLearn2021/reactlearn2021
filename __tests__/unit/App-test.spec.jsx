import { render, screen } from '@testing-library/react';
import renderer from "react-test-renderer";
import LoginRegistrationForm from '../../src/Components/Login/LoginRegistrationForm';
import React from "react";
import LoftTaxi from '../../src/LoftTaxi';
import { shallow } from "enzyme"; // не нужно использовать, так как поддерживается только до 16-ой версии
import { AuthProvider } from "../../src/Components/AuthContext";
import RegistrationForm from "../../src/Components/Login/RegistrationForm";

jest.mock("../../src/assets/logo-auth-left.svg");
jest.mock("mapbox-gl");

test("renders learn react link", () => {
    const { container } = render(<LoginRegistrationForm />);
    expect(container.innerHTML).toMatch("Новый пользователь?");
});

test("renders shallow", () => {
    // const wrapper = shallow(<AuthProvider />);
    // expect(wrapper.contains(<LoftTaxi />)).toBeTruthy(); // не поддерживается версия выше 16-ой
    const wrapper = shallow(<RegistrationForm />);
    expect(wrapper.find("div")).toBeDefined();
    console.info(wrapper.find("div"));
    // wrapper.find("button").simulate("click");
});

test("it renders correctly", () => {
    const tree = renderer.create(<AuthProvider />).toJSON(),
          loftTree = renderer.create(<LoftTaxi />).toJSON(),
          { getByTestId } = render(<LoftTaxi loggedIn={true} />);
    
    expect(tree).toMatchSnapshot(); // отрендерился как null
    expect(loftTree).toMatchSnapshot();
    expect(getByTestId("map-wrapper").firstChild).toBeTruthy(); // потому что loggedIn === true
});
