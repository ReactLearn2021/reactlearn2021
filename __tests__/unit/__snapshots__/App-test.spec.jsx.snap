// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `null`;

exports[`it renders correctly 2`] = `
<main
  data-testid="main"
>
  <section
    id="loginPage"
  >
    <div
      className="left__logo-image"
    >
      <img
        alt="Logotip was not loaded"
        src={
          Object {
            "process": [MockFunction],
          }
        }
      />
    </div>
    <div
      className="auth__section"
    >
      <div
        className="authorize__block"
      >
        <div
          className="authorize__block-form"
        >
          <h2>
            Войти
          </h2>
          <input
            className="loft__form-input"
            id="login"
            name="login"
            onChange={[Function]}
            onFocus={[Function]}
            type="text"
            value=""
          />
           
          <label
            htmlFor="login"
          >
            Имя пользователя 
            <sup>
              ❉
            </sup>
          </label>
          <input
            className="loft__form-input"
            id="password"
            name="password"
            onChange={[Function]}
            onFocus={[Function]}
            type="password"
            value=""
          />
          <label
            htmlFor="password"
          >
            Пароль 
            <sup>
              ❉
            </sup>
          </label>
          <button
            className="loft__form-button"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Войти
          </button>
          <div
            className="loft__form-anchor"
          >
            <span>
              Новый пользователь? 
              <a
                onClick={[Function]}
              >
                Зарегистрироваться
              </a>
            </span>
          </div>
        </div>
      </div>
    </div>
  </section>
</main>
`;
